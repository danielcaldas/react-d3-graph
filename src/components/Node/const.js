export default {
    ARC: {
        START_ANGLE: 0,
        END_ANGLE: 2 * Math.PI
    },
    DEFAULT_NODE_SIZE: 80, // FIXME: This value should come from ../Graph/config.js
    NODE_LABEL_DX: '.90em',
    NODE_LABEL_DY: '.35em',
    SYMBOLS: {
        CIRCLE: 'circle',
        CROSS: 'cross',
        DIAMOND: 'diamond',
        SQUARE: 'square',
        STAR: 'star',
        TRIANGLE: 'triangle',
        WYE: 'wye'
    }
};
