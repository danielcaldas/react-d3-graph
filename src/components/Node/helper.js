/**
 * @module Node/helper
 * @description
 * Some methods that help no the process of rendering a node.
 */
import * as d3 from 'd3';

import CONST from './const';

/**
 * Converts a string that specifies a symbol into a concrete instance
 * of d3 symbol.<br/>
 * {@link https://github.com/d3/d3-shape/blob/master/README.md#symbol}
 * @param  {string} typeName - the string that specifies the symbol type (should be one of {@link #node-symbol-type|node.symbolType}).
 * @return {Object} concrete instance of d3 symbol (defaults to circle).
 * @memberof Node/helper
 */
function _convertTypeToD3Symbol(typeName) {
    switch (typeName) {
        case CONST.SYMBOLS.CIRCLE:
            return d3.symbolCircle;
        case CONST.SYMBOLS.CROSS:
            return d3.symbolCross;
        case CONST.SYMBOLS.DIAMOND:
            return d3.symbolDiamond;
        case CONST.SYMBOLS.SQUARE:
            return d3.symbolSquare;
        case CONST.SYMBOLS.STAR:
            return d3.symbolStar;
        case CONST.SYMBOLS.TRIANGLE:
            return d3.symbolTriangle;
        case CONST.SYMBOLS.WYE:
            return d3.symbolWye;
        default:
            return d3.symbolCircle;
    }
}

/**
 * Build a d3 svg symbol based on passed symbol and symbol type.
 * @param  {number} [size=80] - the size of the symbol.
 * @param  {string} [symbolTypeDesc='circle'] - the string containing the type of symbol that we want to build
 * (should be one of {@link #node-symbol-type|node.symbolType}).
 * @return {Object} concrete instance of d3 symbol.
 * @memberof Node/helper
 */
function buildSvgSymbol(size=80, symbolTypeDesc=CONST.SYMBOLS.CIRCLE) {
    return d3.symbol()
            .size(() => size)
            .type(() => _convertTypeToD3Symbol(symbolTypeDesc))(); // @todo: Strange behavior d3.Symbol ret function
}

export default {
    buildSvgSymbol
};
