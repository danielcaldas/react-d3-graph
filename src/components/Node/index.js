import React from 'react';

import CONST from './const';

import NodeHelper from './helper';

export default class Node extends React.Component {
    // Properties more likely to mutate are evaluated first to take advantage of short-circuit evaluation
    shouldComponentUpdate(nextProps) {
        return nextProps.cx !== this.props.cx || nextProps.cy !== this.props.cy
              || nextProps.opacity !== this.props.opacity
              || nextProps.fill !== this.props.fill
              || nextProps.fontWeight !== this.props.fontWeight
              || nextProps.fontSize !== this.props.fontSize
              || nextProps.label !== this.props.label
              || nextProps.cursor !== this.props.cursor
              || nextProps.size !== this.props.size
              || nextProps.type !== this.props.type
              || nextProps.renderLabel !== this.props.renderLabel
              || nextProps.stroke !== this.props.stroke
              || nextProps.strokeWidth !== this.props.strokeWidth;
    }

    /**
     * Handle click on the node.
     * @return {undefined}
     */
    handleOnClickNode = () => this.props.onClickNode && this.props.onClickNode(this.props.id);

    /**
     * Handle mouse over node event.
     * @return {undefined}
     */
    handleOnMouseOverNode = () => this.props.onMouseOverNode && this.props.onMouseOverNode(this.props.id);

    /**
     * Handle mouse out node event.
     * @return {undefined}
     */
    handleOnMouseOutNode = () => this.props.onMouseOut && this.props.onMouseOut(this.props.id);

    render() {
        const gProps = {
            className: this.props.className,
            cx: this.props.cx,
            cy: this.props.cy,
            id: this.props.id,
            transform: `translate(${this.props.cx},${this.props.cy})`
        };

        const pathProps = {
            cursor: this.props.cursor,
            d: NodeHelper.buildSvgSymbol(this.props.size, this.props.type),
            fill: this.props.fill,
            onClick: this.handleOnClickNode,
            onMouseOut: this.handleOnMouseOutNode,
            onMouseOver: this.handleOnMouseOverNode,
            opacity: this.props.opacity,
            stroke: this.props.stroke,
            strokeWidth: this.props.strokeWidth
        };

        const textProps = {
            dx: CONST.NODE_LABEL_DX,
            dy: CONST.NODE_LABEL_DY,
            fontSize: this.props.fontSize,
            fontWeight: this.props.fontWeight,
            opacity: this.props.opacity
        };

        return (
            <g {...gProps}>
                <path {...pathProps}/>
                {this.props.renderLabel && <text {...textProps}>{this.props.label}</text>}
            </g>
        );
    }
}
