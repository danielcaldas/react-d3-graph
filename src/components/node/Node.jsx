import React from "react";

import nodeHelper from "./node.helper";

/**
 * Node component is responsible for encapsulating node render.
 * @example
 * const onClickNode = function(nodeId) {
 *     window.alert('Clicked node', nodeId);
 * };
 *
 * const onRightClickNode = function(nodeId) {
 *     window.alert('Right clicked node', nodeId);
 * }
 *
 * const onMouseOverNode = function(nodeId) {
 *     window.alert('Mouse over node', nodeId);
 * };
 *
 * const onMouseOutNode = function(nodeId) {
 *     window.alert('Mouse out node', nodeId);
 * };
 *
 * const generateCustomNode(node) {
 *     return <CustomComponent node={node} />;
 * }
 *
 * <Node
 *     id='nodeId'
 *     cx=22
 *     cy=22
 *     fill='green'
 *     fontSize=10
 *     fontColor='black'
 *     fontWeight='normal'
 *     dx=90
 *     label='label text'
 *     labelPosition='top'
 *     opacity=1
 *     renderLabel=true
 *     size=200
 *     stroke='none'
 *     strokeWidth=1.5
 *     svg='assets/my-svg.svg'
 *     type='square'
 *     viewGenerator={generateCustomNode}
 *     className='node'
 *     onClickNode={onClickNode}
 *     onRightClickNode={onRightClickNode}
 *     onMouseOverNode={onMouseOverNode}
 *     onMouseOutNode={onMouseOutNode} />
 */
export default class Node extends React.Component {
    /**
     * Handle click on the node.
     * @returns {undefined}
     */
    handleOnClickNode = () => this.props.onClickNode && this.props.onClickNode(this.props.id);

    /**
     * Handle right click on the node.
     * @param {Object} event - native event.
     * @returns {undefined}
     */
    handleOnRightClickNode = event => this.props.onRightClickNode && this.props.onRightClickNode(event, this.props.id);

    /**
     * Handle mouse over node event.
     * @returns {undefined}
     */
    handleOnMouseOverNode = () => this.props.onMouseOverNode && this.props.onMouseOverNode(this.props.id);

    /**
     * Handle mouse out node event.
     * @returns {undefined}
     */
    handleOnMouseOutNode = () => this.props.onMouseOut && this.props.onMouseOut(this.props.id);

    render() {
        const nodeProps = {
            cursor: this.props.cursor,
            onClick: this.handleOnClickNode,
            onContextMenu: this.handleOnRightClickNode,
            onMouseOut: this.handleOnMouseOutNode,
            onMouseOver: this.handleOnMouseOverNode,
            opacity: this.props.opacity,
        };

        const textProps = {
            ...nodeHelper.getLabelPlacementProps(this.props.dx, this.props.labelPosition),
            fill: this.props.fontColor,
            fontSize: this.props.fontSize,
            fontWeight: this.props.fontWeight,
            opacity: this.props.opacity,
        };

        const size = this.props.size;

        let gtx = this.props.cx,
            gty = this.props.cy,
            label = null,
            node = null;

        if (this.props.svg || this.props.viewGenerator) {
            const height = size / 10;
            const width = size / 10;
            const tx = width / 2;
            const ty = height / 2;
            const transform = `translate(${tx},${ty})`;

            label = (
                <text {...textProps} transform={transform}>
                    {this.props.label}
                </text>
            );

            // By default, if a view generator is set, it takes precedence over any svg image url
            if (this.props.viewGenerator && !this.props.overrideGlobalViewGenerator) {
                node = (
                    <svg {...nodeProps} width={width} height={height}>
                        <foreignObject x="0" y="0" width="100%" height="100%">
                            <section style={{ height, width, backgroundColor: "transparent" }}>
                                {this.props.viewGenerator(this.props)}
                            </section>
                        </foreignObject>
                    </svg>
                );
            } else {
                node = <image {...nodeProps} href={this.props.svg} width={width} height={height} />;
            }

            // svg offset transform regarding svg width/height
            gtx -= tx;
            gty -= ty;
        } else {
            nodeProps.d = nodeHelper.buildSvgSymbol(size, this.props.type);
            nodeProps.fill = this.props.fill;
            nodeProps.stroke = this.props.stroke;
            nodeProps.strokeWidth = this.props.strokeWidth;

            label = <text {...textProps}>{this.props.label}</text>;
            node = <path {...nodeProps} />;
        }

        const gProps = {
            className: this.props.className,
            cx: this.props.cx,
            cy: this.props.cy,
            id: this.props.id,
            transform: `translate(${gtx},${gty})`,
        };

        return (
            <g {...gProps}>
                {node}
                {this.props.renderLabel && label}
            </g>
        );
    }
}
