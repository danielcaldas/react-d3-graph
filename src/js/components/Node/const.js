export default {
    ARC: {
        START_ANGLE: 0,
        END_ANGLE: 2 * Math.PI
    },
    FONT_WEIGHT: 'normal',
    NODE_LABEL_DX: '.90em',
    NODE_LABEL_DY: '.35em',
    SYMBOLS: {
        CIRCLE: 'circle',
        CROSS: 'cross',
        DIAMOND: 'diamond',
        SQUARE: 'square',
        STAR: 'star',
        TRIANGLE: 'triangle',
        WYE: 'wye'
    }
};
