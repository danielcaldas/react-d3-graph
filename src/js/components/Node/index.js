import React from 'react';
import * as d3 from 'd3';

import CONST from './const';
import DEFAULT_CONFIG from '../graph.config';

export default class Node extends React.Component {
    constructor(props) {
        super(props);
    }

    clickNode() {
        console.log('you clicked the node');
    }

    mouseOverNode() {
        console.log('the mouse is over the node');
    }

    render() {
        /*

        const customNodeDrag = d3.drag().on('start', dragstart).on('drag', dragmove).on('end', dragend);

        node = g.selectAll('.node').data(graph.nodes).enter().append('g').attr('class', 'node').call(customNodeDrag);

        if (config.outline) {
            tocolor = 'stroke';
            towhite = 'fill';
        }

        node.append('path')
            .attr('d', buildSvgSymbol(config))
            .style(tocolor, (d) => {
                if (d && d.color) {
                    return d.color;
                }
                return isNumber(d.score) && d.score >= 0 ? color(d.score) : config.defaultNodeColor;
            })
            .style('stroke-width', config.strokeThickness)
            .style(towhite, 'white');

        text = g.selectAll('.text').data(graph.nodes).enter().append('text').attr('dy', '.35em').style('font-size', config.defaultTextSize + 'px');

        if (config.textCenter) {
            text.text((d) => d.id).style('text-anchor', 'middle');
        } else {
            text.attr('dx', (d) => size(d.size) || config.defaultNodeSize)
                .text((d) => d[config.labelProperty] ? '\u2002' + d[config.labelProperty] : '\u2002' + d.id);
        }

        // OUTPUT
        <g class="node" transform="translate(947.7981677565889,254.01685774013785)" cx="947.7981677565889" cy="254.01685774013785">
            <path d="M0,8A8,8 0 1,1 0,-8A8,8 0 1,1 0,8Z" style="fill: rgb(85, 247, 250); stroke-width: 1.5; stroke: white; opacity: 1;">
            </path>
        </g>

        <text dy=".35em" dx="8" transform="translate(947.7981677565889,254.01685774013785)" style="font-size: 9.99778px; font-weight: normal; opacity: 1;">&ensp;100000987214375</text>

         */
        const config = this.props.config;
                // path.arc(x, y, radius, startAngle, endAngle[, anticlockwise])
         /*----------------------------------------
            From above
         ----------------------------------------*/
         const cx = '435.04960405472997';
         const cy = '19.85853539086998';

         const pathX = 0;
         const pathY = 20;
         const radius = config.defaultNodeSize;

         const d = {
             id: 1,
             uid: 'some node'
         };

         const center = false;
         let dx = undefined;
         let labelText = '';

         /*----------------------------------------*/


        const context = d3.path();
        context.arc(pathX, pathY, radius, CONST.ARC.START_ANGLE, CONST.ARC.END_ANGLE);

        let textStyle = {
            fontSize: '10px',
            fontWeight: 'normal',
            opacity: 1
        };

        let pathStyle = {
            fill: 'rgb(85, 247, 250)',
            strokeWidth: 1.5,
            stroke: 'white',
            opacity: 1
        };

        if (center) {
            textStyle['textAnchor'] = 'middle';
        } else {
            // dx = size(d.size) || config.defaultNodeSize;
            // text = g.selectAll('.text').data(graph.nodes).enter().append('text').attr('dy', '.35em').style('font-size', config.defaultTextSize + 'px');
            dx = config.defaultNodeSize;
            labelText = d[config.labelProperty] ? '\u2002' + d[config.labelProperty] : '\u2002' + d.id;
        }

        let pathProps = {
            d: context,
            style: pathStyle
        };

        let textProps = {
            dy: '2.1em',
            dx: dx,
            style: textStyle
        };

        return (
            <g id='1' className='node' transform={`translate(${cx},${cy})`} cx={cx} cy={cy} onClick={this.clickNode} onMouseOver={this.mouseOverNode}>
                <path {...pathProps}/>
                <text {...textProps}>{labelText}</text>
            </g>
        );
    }
}
