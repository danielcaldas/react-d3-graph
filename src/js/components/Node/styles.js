export default {
    textStyle: {
        fontSize: '10px',
        fontWeight: 'normal',
        opacity: 1
    },
    pathStyle: {
        cursor: 'pointer',
        fill: 'rgb(85, 247, 250)',
        strokeWidth: 1.5,
        stroke: 'white',
        opacity: 1
    }
};
