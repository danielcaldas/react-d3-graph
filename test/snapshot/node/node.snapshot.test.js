import React from 'react';
import renderer from 'react-test-renderer';

import Node from '../../../src/components/node/Node';

describe('Snapshot - Node Component', () => {
    let that = {};

    beforeEach(() => {
        that.clickCallback = jest.fn();
        that.mouseOverCallback = jest.fn();
        that.mouseOutCallback = jest.fn();

        that.node = renderer.create(
            <Node
                x1="2"
                y1="2"
                x2="4"
                y2="4"
                opacity="1"
                fill="black"
                fontColor="black"
                fontWeight="bold"
                fontSize="12"
                label="text"
                cursor="pointer"
                size="120"
                type="circle"
                renderLabel="true"
                stroke="red"
                strokeWidth="2"
                onClickNode={that.clickCallback}
                onMouseOverNode={that.mouseOverCallback}
                onMouseOut={that.mouseOutCallback}
            />
        );

        that.tree = that.node.toJSON();
    });

    test('should match snapshot', () => {
        expect(that.tree).toMatchSnapshot();
    });
});
