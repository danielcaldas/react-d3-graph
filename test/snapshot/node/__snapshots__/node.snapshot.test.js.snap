// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot - Node Component should match snapshot 1`] = `
<g
  className={undefined}
  cx={undefined}
  cy={undefined}
  id={undefined}
  transform="translate(undefined,undefined)"
>
  <path
    cursor="pointer"
    d="M6.180387232371033,0A6.180387232371033,6.180387232371033,0,1,1,-6.180387232371033,0A6.180387232371033,6.180387232371033,0,1,1,6.180387232371033,0"
    fill="black"
    onClick={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    opacity="1"
    stroke="red"
    strokeWidth="2"
  />
  <text
    dx=".90em"
    dy=".35em"
    fill="black"
    fontSize="12"
    fontWeight="bold"
    opacity="1"
  >
    text
  </text>
</g>
`;
